OPHD/StructureManager.h

```diff
@@ -52,6 +53,79 @@ class StructureManager
 
 	void serialize(NAS2D::Xml::XmlElement* element);
 
+	/**
+	 * Associate a StructureComponent with a structure.
+	 * This transfers ownership of the component instance to the
+	 * structure manager. The component will be destroyed when the
+	 * structure with which it is associated is removed.
+	 */
+	template<typename ComponentTy>
+	void attachComponent(SKey s, ComponentTy* component)
+	{
+		auto& table = mComponents[ComponentTy::componentTypeID];
+		bool success = table.insert(std::make_pair(s, static_cast<StructureComponent*>(component))).second;
+#if defined(_DEBUG)
+		if (!success)
+		{
+			throw std::runtime_error("Structure::Attach() was called on a Structure that already had the component!");
+		}
+#endif
+	}
+
+	/**
+	 * Return a reference to the given StructureComponent type belonging to
+	 * a structure. The structure is assumed to have the given component,
+	 * and it is an error to try to get a component from a structure that
+	 * does not have it.
+	 */
+	template<typename ComponentTy>
+	ComponentTy& get(SKey s)
+	{
+		ComponentTy* component = tryGet<ComponentTy>(s);
+#if defined(_DEBUG)
+		if (!component)
+		{
+			throw std::runtime_error("StructureManager::get() was called on a Structure without the requested component!");
+		}
+#endif
```

Comment:
This function does not try to do both. This function returns a reference, and it is an error to call it for a structure that does not have the component. Only throwing in debug builds was mimicking what several other nearby locations in the code were already doing. This is actually a third form which is very common in release builds: programmer errors cause the application to crash. But I can make the error check unconditional.

The tryGet function is of form (1), distingueshed by function name like you mentioned is typical. That one is allowed to fail, as implied by its name.

Reply:
Understood about mimicking other parts of the code. Those parts of the code were written with a different understanding of how to effectively implement C++ debug handling from long, long ago. I've been slowly removing those kinds of things.

Anyway, it's not a bad way to do it, just not the preferred method any longer. I'm mostly concerned about readable and maintainable code that behaves the same way in debug and release builds. Having a throw in a release build is still helpful because it can help us sort mistakes that are otherwise not found during typical testing.

Reply:
I did a search and did find a few bits of code that are still using `#if defined`/`#if def`.
 - Factory.cpp
 - Structure.cpp
 - Thing.h
We should probably update those at some point.

----

You're quite right that `tryGet` uses form (1). Then this function tries to convert from form (1) to form (2), but only for debug builds. That's why I said this function tries to handle errors in both ways. By virtue of delegating to `tryGet`, this function uses form (1) in release mode, and form (2) in debug mode.

Form (3) also makes sense, and I guess that's kind of what you were going for. I guess I normally expect a complete lack of error checking or reporting code for form (3).

I'm probably getting a bit pedantic here though.

Reply:
Fair. My day job is in GPU driver dev, so I'm used to hunting microseconds of execution time in release builds and using separate builds for development and debugging purposes. But I'm happy to throw in release now that I know your philosophy.